src/main/antlr/GroovyParser.g4

```diff
@@ -347,14 +347,13 @@ options { baseContext = type; }
         dimsOpt
     ;
 
-type
+type[boolean allowVoid]
```

Comment:
:bulb: I like how this is used in the visitor instead of relying on semantic predicates. Consider adding a comment here explaining that the argument is intended for use after the parse is complete.

❔ I don't see any instance of `type[true]` in the grammar.

:bulb: It's not clear to me if the goal here is allowing `type` to be used in cases where VOID is allowed and also VOID isn't allowed (a simplification of the grammar and parse tree), or if the goal is to defer warnings related to improper use of VOID until after the parse is complete (potentially a big usability benefit to error recovery if incorrect placement of VOID is something users are known to do). If the *main goal* is ability to reuse the `TypeContext` type in later compiler passes, another option may be this:

```antlr
typeNoVoid
options { baseContext = type; }
  : annotationsOpt
    ( primitiveType
    | generalClassOrInterfaceType
    )
    dimsOpt
  ;
```

Reply:
@sharwell `type[true]` is one part of an experimental solution, which is eventually given up... I'll refine it later. As for invalid `void`, e.g. `void x = 123`, friendly error messages are wanted, this is why I did not apply some rule like `typeNoVoid`. Thanks for your reviewing the code ;-)